fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,180 @@
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+public class LuceneSerializer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSerializer.class);
+    private static final int ENCODING_ENCRYPTED = 1;
+    private static final int ENCODING_COMPRESSED = 2;
+    private static final byte COMPRESSION_VERSION_FOR_HIGH_COMPRESSION = 0;
+
+    private static class EncodingState {
+        private boolean compressed;
+        private boolean encrypted;
+
+        private EncodingState() {
+            this.compressed = false;
+            this.encrypted = false;
+        }
+
+        private void setCompressed(boolean compressed) {
+            this.compressed = compressed;
+        }
+
+        private void setEncrypted(boolean encrypted) {
+            this.encrypted = encrypted;
+        }
+
+        private boolean isCompressed() {
+            return compressed;
+        }
+
+        private boolean isEncrypted() {
+            return encrypted;
+        }
+    }
+
+    @Nullable
+    public static byte[] encode(@Nullable byte[] data, boolean compress, boolean encrypt) {
+        if (data == null) {
+            return null;
+        }
+
+        final EncodingState state = new EncodingState();
+        final ByteBuffersDataOutput decodedDataOutput = new ByteBuffersDataOutput();
+        // Placeholder for the code byte at beginning, will be modified in output byte array if needed
+        decodedDataOutput.writeByte((byte) 0);
+
+        byte[] encoded = compressIfNeeded(compress, data, decodedDataOutput, state, 1);
+
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        // Encryption will be supported in future
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        LOGGER.trace("Encoded data for compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+
+        byte[] decoded = decompressIfNeeded(compressed, data, 1);
+        LOGGER.trace("Decoded data for compressed={}, encrypted={}, encodedDataSize={}, decodedDataSize={}",
```

Comment:
Why is it using this style of logging instead of using a `KeyValueLogMessage` (possibly guarded by an `if` block conditional on `LOGGER.isTraceEnabled()`)?

Reply:
Makes sense. Changed to ```KeyValueLogMessage``` for the places touched by this PR. I guess the other loggings in ```FDBDirectory``` may also need to be changed to ```KeyValueLogMessage```, but in a separate PR probably.

Reply:
Yeah, that sounds right. We do want to have `KeyValueLogMessage` as our standard interface to logs everywhere, but that larger refactoring can definitely be a separate change from this one